sw/device/lib/dif/dif_rv_timer.h

```diff
@@ -0,0 +1,296 @@
+  kDifRvTimerEnabled,
+} dif_rv_timer_enabled_t;
+
+/**
+ * Represents timekeeping parameters for a particular timer.
+ */
+typedef struct dif_rv_timer_tick_params {
+  /**
+   * The perscaler value, which is a unitless conversion from clock cycles
+   * to the desired timer tick frequency, minus one. That is,
+   *
+   *   prescale = (clock_freq * tick_length) - 1
+   *
+   * with |clock_freq| and |tick_length| given in units of hertz and seconds,
+   * respectively.
+   *
+   * For example, if the clock frequency is 50 MHz, and the desired tick
+   * length is is 1 MHz (i.e., 1 microsecond ticks), this value should be
+   * set to |50/1 - 1 = 49|.
+   *
+   * This value is declared as a uint16_t, but only the lowest 12 bits are
+   * actually used.
+   */
+  uint16_t prescale;
+
+  /**
+   * The amount to step the timer counter by each tick.
+   */
+  uint8_t tick_step;
+} dif_rv_timer_tick_params_t;
+
+/**
+ * Errors returned by |dif_rv_timer_init()|.
+ */
+typedef enum dif_rv_timer_approximate_tick_params_result {
+  /**
+   * Indicates a successful operation.
+   */
+  kDifRvTimerApproximateTickParamsOk = 0,
+  /**
+   * Indicates a failed precondition on the function arguments.
+   */
+  kDifRvTimerApproximateTickParamsPreconditionFailed,
+  /**
+   * Indicates that the requested counter frequency cannot be represented
+   * without loss in precission.
+   */
+  kDifRvTimerApproximateTickParamsUnrepresentable,
+} dif_rv_timer_approximate_tick_params_result_t;
+
+/**
+ * Generates an aproximate |dif_rv_timer_tick_params_t| given the device
+ * clock frequency and desired counter frequency (both given in Hertz).
+ *
+ * For the purposes of this function, "counter frequency" is the frequency
+ * at which software would observe a timer counter to increase. If the
+ * clock has insufficient resolution, high counter frequencies may set a
+ * larger value for |tick_step|. For example, if the clock ticks at 50kHz,
+ * but we want a counter that seems to tick every microsecond (1MHz),
+ * we can achieve this with a prescale of 0 (so that there is a tick per
+ * clock cycle) and a tick step of 20 (since 20 * 50kHz = 1MHz).
+ *
+ * The return value of this function is only an approximation, and the
+ * actual counter frequency ultimately depends on the accuracy of the
+ * clock.
+ *
+ * @param clock_freq the device clock frequency, in Hertz.
+ * @param counter_freq the desired counter frequency, in Hertz.
+ * @param out tick parameters that will approximately produce the desired
+ *         counter frequency.
+ * @return the result of the operation.
+ */
+dif_rv_timer_approximate_tick_params_result_t
+dif_rv_timer_approximate_tick_params(uint64_t clock_freq, uint64_t counter_freq,
+                                     dif_rv_timer_tick_params_t *out);
+
+/**
+ * Configuration for initializing a RISC-V timer.
+ */
+typedef struct dif_rv_timer_config {
+  /**
+   * The physical location of the timer in memory.
+   */
+  mmio_region_t base_addr;
```

Comment:
Just to check my understanding: This will be the same for all timers, right? (`CTRL` register is at this address, and the rest of the offsets depend on the hart index.)

Reply:
I *think* so, yes. There is only one timer device but there can be many harts sharing it.